research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+  else:
+    learned_whitening = None
+
+  # Evaluate on test datasets.
+  for dataset in datasets:
+    start = time.time()
+
+    # Prepare config structure for the test dataset.
+    cfg = test_dataset.CreateConfigForTestDataset(dataset,
+                                                  os.path.join(data_root))
+    images = [cfg['im_fname'](cfg, i) for i in range(cfg['n'])]
+    qimages = [cfg['qim_fname'](cfg, i) for i in range(cfg['nq'])]
+    bounding_boxes = [tuple(cfg['gnd'][i]['bbx']) for i in range(cfg['nq'])]
+
+    # Extract database and query vectors.
+    global_features_utils.debug_and_log(
+            '>> {}: Extracting database images...'.format(dataset))
+    vecs = global_model.extract_global_descriptors_from_list(
+            net, images, image_size, ms=multiscale)
+    global_features_utils.debug_and_log(
+            '>> {}: Extracting query images...'.format(dataset))
+    qvecs = global_model.extract_global_descriptors_from_list(
+            net, qimages, image_size, bounding_boxes,
+            ms=multiscale)
+
+    global_features_utils.debug_and_log('>> {}: Evaluating...'.format(dataset))
+
+    # Convert the obtained descriptors to numpy.
+    vecs = vecs.numpy()
+    qvecs = qvecs.numpy()
+
+    # Search, rank and print test set metrics.
+    scores = np.dot(vecs.T, qvecs)
+    ranks = np.transpose(np.argsort(-scores, axis=0))
+    metrics = global_features_utils.compute_metrics_and_print(dataset, ranks,
+                                                              cfg['gnd'])
+    # Save calculated metrics in a tensorboard format.
+    if writer:
+      tf.summary.scalar('test_accuracy_{}_E'.format(dataset), metrics[0][0],
+                        step=epoch)
+      tf.summary.scalar('test_accuracy_{}_M'.format(dataset), metrics[1][0],
+                        step=epoch)
+      tf.summary.scalar('test_accuracy_{}_H'.format(dataset), metrics[2][0],
+                        step=epoch)
+      writer.flush()
+
+    if learned_whitening is not None:
+
+      # Whiten the vectors.
+      mean_vector = learned_whitening['m']
+      projection_matrix = learned_whitening['P']
+      vecs_lw = whiten.whitenapply(vecs, mean_vector, projection_matrix)
+      qvecs_lw = whiten.whitenapply(qvecs, mean_vector, projection_matrix)
+
+      # Search, rank, and print.
+      scores = np.dot(vecs_lw.T, qvecs_lw)
+      ranks = np.transpose(np.argsort(-scores, axis=0))
+      metrics = global_features_utils.compute_metrics_and_print(
+              dataset + ' + whiten', ranks, cfg['gnd'])
+
+      if writer:
+        tf.summary.scalar('test_accuracy_whiten_{}_E'.format(dataset),
+                          metrics[0][0], step=epoch)
+        tf.summary.scalar('test_accuracy_whiten_{}_M'.format(dataset),
+                          metrics[1][0], step=epoch)
+        tf.summary.scalar('test_accuracy_whiten_{}_H'.format(dataset),
+                          metrics[2][0], step=epoch)
+        writer.flush()
```

Comment:
seems basically duplicated code from 328-341. Maybe we could create a small auxiliary function to avoid this / simplify things here.